code/ship/ship.cpp

```diff
@@ -12636,23 +12636,21 @@ int ship_fire_primary(object * obj, int force, bool rollback_shot)
 				}
 			}
 			
-			if(winfo_p->wi_flags[Weapon::Info_Flags::Beam]){		// the big change I made for fighter beams, if there beams fill out the Fire_Info for a targeting laser then fire it, for each point in the weapon bank -Bobboau
-				beam_fire_info fbfire_info;				
+			if(winfo_p->wi_flags[Weapon::Info_Flags::Beam]){		// the big change I made for fighter beams, if there beams fill out the Fire_Info for a targeting laser then fire it, for each point in the weapon bank -Bobboau				
 
-				int points;
+				int points = 0, numtimes = 1;
 				if (winfo_p->b_info.beam_shots){
-					if (winfo_p->b_info.beam_shots > num_slots){
-						points = num_slots;
-					}else{
-						points = winfo_p->b_info.beam_shots;
-					}
+					numtimes = winfo_p->shots;
+					points = MIN(winfo_p->b_info.beam_shots, num_slots);
 				} else if (winfo_p->wi_flags[Weapon::Info_Flags::Cycle]) {
-					points = 1;
+					numtimes = 1;
+					points = MIN(num_slots, winfo_p->shots);
```

Comment:
Is this the right thing to do here?

Reply:
According to the current way "cycle" works, yes. With "cycle" enabled, $Shots doesn't change the number of shots generated by an individual firing event; instead, it controls the number of firepoints from which the weapon fires. So you can have a bank with 4 firepoints, and a weapon with a $Shots of 3 and the "cycle" flag, and what'll happen is that when the weapon fires, it'll fire from 3 firepoints, then cycle to the next set of three on the next shot.

I think that this is maybe not a great way to handle the behavior and we should possibly think about changing it, but that's beyond the scope of this feature.